packages/neon_framework/packages/notifications_push_repository/lib/src/notifications_push_repository.dart

```diff
@@ -0,0 +1,243 @@
+      _previousDistributor = _selectedDistributor;
+      _log.finer('UnifiedPush distributor changed to $_selectedDistributor');
+
+      await _unregisterUnifiedPush();
+      await UnifiedPush.saveDistributor(_selectedDistributor!);
+    }
+
+    await _registerUnifiedPush();
+  }
+
+  Future<void> _registerUnifiedPush() async {
+    // Notifications will only work on accounts with app password
+    for (final account in _accounts.where((a) => a.password != null)) {
+      _log.finer('Registering ${account.id} for UnifiedPush');
+
+      await UnifiedPush.registerApp(account.id);
+    }
+  }
+
+  Future<void> _unregisterUnifiedPush() async {
+    final subscriptions = await _storage.readSubscriptions();
+    for (final entry in subscriptions.entries) {
+      final accountID = entry.key;
+      final account = _accounts.firstWhereOrNull((account) => account.id == accountID);
+      var subscription = entry.value;
+
+      subscription = await _unregisterNextcloud(accountID, account, subscription);
+      if (subscription.endpoint != null) {
+        _log.finer('Unregistering $accountID from UnifiedPush');
+
+        subscription = subscription.rebuild((b) => b.endpoint = null);
+        await _saveUpdatedSubscription(subscriptions, accountID, subscription);
+
+        await UnifiedPush.unregister(accountID);
+      } else {
+        await _saveUpdatedSubscription(subscriptions, accountID, subscription);
+      }
+    }
+  }
+
+  Future<PushSubscription> _registerNextcloud(Account account, String endpoint, PushSubscription subscription) async {
+    _log.finer('Registering ${account.id} at Nextcloud');
+
+    try {
+      final response = await account.client.notifications.push.registerDevice(
+        $body: notifications.PushRegisterDeviceRequestApplicationJson(
+          (b) => b
+            ..pushTokenHash = notifications.generatePushTokenHash(endpoint)
+            ..devicePublicKey = _privateKey.publicKey.toFormattedPEM()
+            ..proxyServer = '$endpoint#',
+        ),
+      );
+
+      return subscription.rebuild(
+        (b) => b
+          ..endpoint = endpoint
+          ..pushDevice.replace(response.body.ocs.data),
+      );
+    } on DynamiteApiException catch (error) {
+      _log.warning('Failed to register ${account.id} at Nextcloud', error);
+    }
+
+    return subscription;
+  }
+
+  Future<PushSubscription> _unregisterNextcloud(
+    String accountID,
+    Account? account,
+    PushSubscription subscription,
+  ) async {
+    if (subscription.pushDevice != null) {
+      _log.finer('Unregistering $accountID from Nextcloud');
+
+      try {
+        await account?.client.notifications.push.removeDevice();
+      } on DynamiteApiException catch (error) {
+        _log.warning('Failed to unregister $accountID at Nextcloud', error);
+      }
+    }
+
+    // Remove the push device either way, as we need to register for a new one later
+    return subscription.rebuild((b) => b.pushDevice = null);
```

Comment:
Why do we still keep the endpoint?
What value does it have to us?

If we don't need it, just unregister from the NC endpoint and remove the subscription entirely.

Reply:
Because it is separate from the Nextcloud subscription.
The endpoint can still be valid while we had to remove/update the Nextcloud subscription.